README.md
# Enhance [![Build Status](https://secure.travis-ci.org/jico/enhance.png?branch=master)](http://travis-ci.org/jico/enhance)

A flexible node.js module for serving high-res images to high pixel density devices.

## Installation

Add `enhance` to your `package.json` file and run `npm install`.

Alternatively, you can run `npm install enhance --save`.

If you're using Hem, make sure to add it to your `slug.json` file as a
dependency.

## Usage

Require _Enhance_ in your app, optionally passing in initialization options, and
typically set to a global variable. In CoffeeScript, it would look something
like:

```coffee
Enhance = require('enhance')(options)
```

### `render`

Given no [configuration options](#configuration), _Enhance_ falls back to
Apple's convention of suffixing image file names with `@2x` before the
extension if a Retina or other high DPI display is detected.

For example, consider calling the main function `render` on a Retina device:

```coffee
Enhance.render('image.png')
# => "image@2x.png"
```

If the device does not have a high DPI, the call to `render` will simply return
the file name unaltered (unless initialization options have been passed).

### `isHiDPI`

You can use the available method `isHiDPI` to check if a device is one of high
DPI. The default device pixel ratio is 1.3, but `isHiDPI` takes an optional
ratio parameter to use as a threshold.

### `isMobileDevice`

To detect whether a user is on a mobile device, you can call `isMobileDevice`,
which uses `max-width` media queries as a test.

By default, tablets are not considered mobile devices. However, you can consider
them by enabling the `tabletAsMobile` option when initializing _Enhance_:

Additionally, you can override the width breakpoints for both mobile and tablet
devices, see [configuration](#configuration). The default breakpoints are:

```coffee
phoneBreakpoint:  480
tabletBreakpoint: 1024
```

## Configuration

_Enhance_ accepts initial configuration options to provide flexibility if your
app doesn't follow Apple's Retina naming convention.

### host

Prepends the given `host` value to file names passed to the call to `render`.

__Default:__ _null_

```coffee
Enhance = require('enhance')(host: 'http://www.example.com')
Enhance.render('image.png')
# => "http://www.example.com/image@2x.png" (high DPI)
```

### suffix

The string to prepend before the file name extension.

__Default:__ _"@2x"_

```coffee
Enhance = require('enhance')(suffix: '_2x')
Enhance.render('image.png')
# => "image_2x.png" (high DPI)
```

### render

Custom callback function to use in place of the default `render` method. Passes
in a helper object that makes available the following:

* `isHiDPI`
* `_` - [lodash](http://lodash.com) function with additional extensions:
  * `joinURIComponents` - Joins any number of URI string component arguments
  while preserving double slashes (i.e. the case of `http://`)
* `src` - String argument of `render`
* Any and all other parameters passed upon initialization or the call to `render`

```coffee
options =
  host: 'http://example.com/transform'
  suffix: '_2x'
  render: (helpers) ->
    if helpers.isHiDPI()
      path = "#{encodeURIComponent(helpers.src)}/resize/#{helpers.width*2}x#{helpers.height*2}#"
    else
      path = "#{encodeURIComponent(helpers.src)}/resize/#{helpers.width}x#{helpers.height}#"
    helpers._.joinURIComponents(helpers.host, path)

Enhance = require('enhance')(options)

Enhance.render('image.png', width: 50, height: 100)
```

### phoneBreakpoint

Breakpoint for detecting mobile phone devices. Used by `isMobileDevice` method.

__Default__: _480_

### tabletBreakpoint

Breakpoint for detecting mobile tablet devices. Used by `isMobileDevice` method.

__Default__: _1024_

### tabletAsMobile

__Default__: _false_

Enables consideration on tablets as mobile devices.


## License

Copyright 2013 Jico Baligod.

Licensed under the [MIT License](http://github.com/jico/enhance/raw/master/LICENSE).


CHANGELOG.md
# Changelog

## v0.0.6 - 2013-10-07

* Preserve relative protocol URIs.

## v0.0.5 - 2013-10-02

* Handle leading and trailing slashes in URI components appropriately (preserves
  absolute URLs).

## v0.0.4 - 2013-10-02

* Registers package on Bower.

## v0.0.3 - 2013-08-12

* Adds `isMobileDevice` method.
* Adds `phoneBreakpoint`, `tabletBreakpoint`, and `tabletAsMobile` options.

## v0.0.2 - 2013-08-11

* Switch to lodash compatibility build for IE8.


test/enhance.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Enhance, expect, fs, jsdom;

  expect = require('expect.js');

  fs = require('fs');

  jsdom = require('jsdom');

  Enhance = require('../lib/enhance');

  describe('Enhance', function() {
    var setNonRetina, setRetina;
    setRetina = function() {
      return window.devicePixelRatio = 2.0;
    };
    setNonRetina = function() {
      return window.devicePixelRatio = 1.0;
    };
    beforeEach(function() {
      window = jsdom.createWindow();
      return window.matchMedia = function() { return { matches: false } };
    });
    describe('init', function() {
      describe('suffix option', function() {
        return it('appends the given suffix for retina devices', function(done) {
          setRetina();
          expect(Enhance({
            suffix: '_2x'
          }).render('image.png')).to.be('image_2x.png');
          return done();
        });
      });
      describe('host option', function() {
        it('prepends the given host for non-retina devices', function(done) {
          var enhance;
          setNonRetina();
          enhance = Enhance({
            host: 'http://example.com'
          }).render('image.png');
          expect(enhance).to.be('http://example.com/image.png');
          return done();
        });
        return it('prepends the given host for retina devices', function(done) {
          var enhance;
          setRetina();
          enhance = Enhance({
            host: 'http://example.com'
          }).render('image.png');
          expect(enhance).to.be('http://example.com/image@2x.png');
          return done();
        });
      });
      return describe('render callback', function() {
        it('calls the passed render callback instead of the default', function(done) {
          var enhance;
          enhance = Enhance({
            render: function(properties) {
              return "awesome";
            }
          }).render('image.png');
          expect(enhance).to.be('awesome');
          return done();
        });
        return describe('the callback enhance object', function(done) {
          it('has the src filename being rendered', function(done) {
            var initOpts;
            initOpts = {
              render: function(enhance) {
                expect(enhance.src).to.be('image.png');
                return done();
              }
            };
            return Enhance(initOpts).render('image.png');
          });
          it('has any enhance passed in at initialization', function(done) {
            var initOpts;
            initOpts = {
              render: function(enhance) {
                expect(enhance.host).to.be('http://example.com');
                expect(enhance.suffix).to.be('_2x');
                return done();
              },
              host: 'http://example.com',
              suffix: '_2x'
            };
            return Enhance(initOpts).render('image.png');
          });
          it('has any properties passed in on call to render', function(done) {
            var initOpts;
            initOpts = {
              render: function(enhance) {
                expect(enhance.width).to.be(50);
                expect(enhance.height).to.be(100);
                return done();
              }
            };
            return Enhance(initOpts).render('image.png', {
              width: 50,
              height: 100
            });
          });
          return it('has helper methods available', function(done) {
            var initOpts;
            initOpts = {
              render: function(enhance) {
                expect(enhance._).to.be.a('function');
                expect(enhance.isHiDPI).to.be.a('function');
                expect(enhance.prependHost).to.be.a('function');
                return done();
              }
            };
            return Enhance(initOpts).render('image.png');
          });
        });
      });
    });
    describe('#isHiDPI', function() {
      it('returns true if devicePixelRatio greater than 1.3 by default', function(done) {
        setRetina();
        expect(Enhance().isHiDPI()).to.be(true);
        return done();
      });
      it('return true if matchMedia has matches on the media query', function(done) {
        setNonRetina();
        window.matchMedia = function() { return { matches: true } };
        expect(Enhance().isHiDPI()).to.be(true);
        return done();
      });
      it('returns false if neither devicePixelRatio nor matchMedia test passes', function(done) {
        window.devicePixelRatio = 1.0;
        window.matchMedia = undefined;
        expect(Enhance().isHiDPI()).to.be(false);
        return done();
      });
      return it('accepts an optional ratio parameter', function(done) {
        window.devicePixelRatio = 1.5;
        expect(Enhance().isHiDPI(1.0)).to.be(true);
        expect(Enhance().isHiDPI(2.0)).to.be(false);
        return done();
      });
    });
    return describe('#render', function() {
      return describe('default Apple Retina naming convention', function() {
        it('returns the original filename for non-retina devices', function(done) {
          expect(Enhance().render('image.png')).to.be('image.png');
          return done();
        });
        return it('returns a suffixed filename for retina devices', function(done) {
          setRetina();
          expect(Enhance().render('image.png')).to.be('image@2x.png');
          return done();
        });
      });
    });
  });

}).call(this);


lib/lodash-ext.js
// Generated by CoffeeScript 1.6.3
(function() {
  var _;

  _ = require('../vendor/lodash/dist/lodash');

  _.mixin({
    joinURIComponents: function() {
      var components, leadingSlashes, uri, _ref;
      leadingSlashes = ((_ref = arguments[0]) != null ? _ref.match(/^[\/]{1,2}/) : void 0) || '';
      components = _.map(arguments, function(str, i) {
        if (str != null) {
          return _.trim(str, '/').split(/\/(?!\/)/);
        }
      });
      return uri = leadingSlashes + _.flatten(_.compact(components)).join('/');
    },
    trim: function(string, characters) {
      var regexp;
      if (characters == null) {
        characters = '\\s';
      }
      regexp = new RegExp("^[" + characters + "]+|[" + characters + "]+$", 'g');
      return string.replace(regexp, '');
    }
  });

  module.exports = _;

}).call(this);


lib/enhance.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Enhance, _;

  _ = require('./lodash-ext');

  Enhance = (function() {
    return function(options) {
      var defaults, exports, helpers, isHiDPI, isMobileDevice, render;
      defaults = {
        host: null,
        suffix: '@2x',
        phoneBreakpoint: 480,
        tabletBreakpoint: 1024,
        tabletAsMobile: false
      };
      options = _.merge(defaults, options);
      isHiDPI = function(ratio) {
        var query;
        if (ratio == null) {
          ratio = 1.3;
        }
        query = "only screen and (-moz-min-device-pixel-ratio: " + ratio + "),         only screen and (-o-min-device-pixel-ratio: " + (ratio * 2) + "/2),         only screen and (-webkit-min-device-pixel-ratio: " + ratio + "),         only screen and (min-device-pixel-ratio: " + ratio + "),         only screen and (min-resolution: " + ratio + "dppx)";
        if (window.devicePixelRatio > ratio) {
          return true;
        }
        if (typeof window.matchMedia === "function" ? window.matchMedia(query).matches : void 0) {
          return true;
        }
        return false;
      };
      isMobileDevice = function() {
        var breakpoint, mobileDevice;
        mobileDevice = options.tabletAsMobile ? 'tablet' : 'phone';
        breakpoint = options["" + mobileDevice + "Breakpoint"];
        return typeof window.matchMedia === "function" ? window.matchMedia("only screen and (max-width: " + breakpoint + "px)").matches : void 0;
      };
      helpers = {
        _: _,
        isHiDPI: isHiDPI
      };
      render = function(src, opts) {
        var regexp;
        if (options.render != null) {
          return typeof options.render === "function" ? options.render(_.merge({
            src: src
          }, options, opts, helpers)) : void 0;
        } else {
          regexp = new RegExp(/(.jpe?g|.png|.gif|.ti?ff?)/i);
          if (isHiDPI()) {
            src = src.replace(regexp, "" + options.suffix + "$1");
          }
          return _.joinURIComponents(options.host, src);
        }
      };
      return exports = {
        isHiDPI: isHiDPI,
        render: render,
        isMobileDevice: isMobileDevice
      };
    };
  })();

  if (typeof window !== "undefined" && window !== null) {
    window.Enhance = Enhance;
  }

  if (typeof module !== "undefined" && module !== null) {
    module.exports = Enhance;
  }

}).call(this);
